### PR TITLE
LSIF: Replace integration tests with unit tests

### PR DESCRIPTION
**Pull request by efritz**
Refactoring steps taken:

- Move the cache instances from the backend to static properties of the database.
- Replace createDatabase instance function of Backend with a function (optionally) supplied in the constructor.
- Move functions unused by the database out of database.ts.
- Simplify some conditions in the backend.
- Remove all integration tests.
- Move non-integration tests (but db-interfacing tests) next to the code they test.
- Add unit tests for backend and database.

This will make it possible to retain the same test functionality after splitting the backend and the database classes into different services.

**Note to reviewers**: These tests are not perfect (no code is), but it's going to unblock us from progressing on [RFC 127](https://docs.google.com/document/d/1lno7N_gDg2RmvI4cLZ7l_TdnxP1e-P1Ri7Iwr2RCjKQ), and the tests may need to change a great deal once we've refactored into services. Do not look _too_ carefully at the test implementations, which are still a good incremental upgrade from the previous state.

### CODE REVIEW

File: lsif/src/server/backend/database.test.ts [hunk 1]
```diff
@@ -1,5 +1,252 @@
...
+
+    afterAll(async () => {
+        if (storageRoot) {
+            await rmfr(storageRoot)
+        }
+    })
+
+    describe('exists', () => {
+        it('should check document path', async () => {
+            expect(await database.exists('cmd/lsif-go/main.go')).toBeTruthy()
```
Comment (creachadair) [inline lsif/src/server/backend/database.test.ts]: Shouldn't a predicate return actual-true?

File: lsif/src/server/backend/database.test.ts [hunk 2]
```diff
@@ -1,5 +1,252 @@
...
+            expect(await database.exists('internal/index/indexer.go')).toBeTruthy()
+            expect(await database.exists('missing.go')).toBeFalsy()
+        })
+    })
+
+    describe('definitions', () => {
+        it('should correlate definitions', async () => {
+            // `\ts, err := indexer.Index()` -> `\t Index() (*Stats, error)`
```
Comment (creachadair) [inline lsif/src/server/backend/database.test.ts]: These example comments are great.

File: lsif/src/server/backend/location.ts [hunk 3]
```diff
@@ -20,6 +20,22 @@ export class OrderedLocationSet {
     private seen = new Set<string>()
     private order: InternalLocation[] = []
 
+    constructor(locations?: InternalLocation[], sorted = false) {
```
Comment (creachadair) [inline lsif/src/server/backend/location.ts]: Can you please add a comment as to the semantics of the `sorted` flag? I reconstructed it by reading the code, but ideally the caller shouldn't have to do that.

File: lsif/src/server/backend/backend.test.ts [hunk 4]
```diff
@@ -0,0 +1,689 @@
...
+    ...zeroDocument,
+    monikers: new Map([
+        [50, { kind: lsif.MonikerKind.local, scheme: 'test', identifier: 'm1' }],
+        [51, { kind: lsif.MonikerKind.import, scheme: 'test', identifier: 'm2', packageInformationId: 71 }],
+        [52, { kind: lsif.MonikerKind.import, scheme: 'test', identifier: 'm3' }],
+    ]),
+    packageInformation: new Map([[71, { name: 'pkg2', version: '0.0.1' }]]),
+}
+
+const ranges = [
```
Comment (creachadair) [inline lsif/src/server/backend/backend.test.ts]: Can this get a comment as to what this range is for? (Is it still "these ranges are" when there's only one? 🙂)

Or possibly a more descriptive name if that's easier.

File: lsif/src/server/backend/location.ts (hunk 3 above)
Comment (efritz) [reply lsif/src/server/backend/location.ts]: Yeah this was to support tests but I'll make sure to do that.

File: lsif/src/server/backend/backend.test.ts (hunk 4 above)
Comment (efritz) [reply lsif/src/server/backend/backend.test.ts]: Np

File: lsif/src/server/backend/database.test.ts (hunk 1 above)
Comment (efritz) [reply lsif/src/server/backend/database.test.ts]: Yup but there's no "toBeTrue".

Comment (creachadair) [reply lsif/src/server/backend/database.test.ts]: Of course there isn't. 🙄

File: lsif/src/server/backend/location.ts [hunk 5]
```diff
@@ -20,6 +20,28 @@ export class OrderedLocationSet {
     private seen = new Set<string>()
     private order: InternalLocation[] = []
 
+    /**
+     * Create a new ordered location set.
+     *
+     * @param locations A set of locations used to seed the set.
+     * @param sorted Whether the given locations are already deduplicated.
```
Comment (creachadair) [inline lsif/src/server/backend/location.ts]: `sorted` → `trusted`

Comment (efritz) [reply lsif/src/server/backend/location.ts]: Durh.

**Pull request merged**